src/v/kafka/server/audit_log_manager.h

```diff
@@ -0,0 +1,156 @@
+// Copyright 2023 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#pragma once
+#include "cluster/fwd.h"
+#include "config/property.h"
+#include "kafka/client/fwd.h"
+#include "kafka/client/types.h"
+#include "kafka/server/logger.h"
+#include "kafka/types.h"
+#include "model/timeout_clock.h"
+#include "net/types.h"
+#include "security/audit/schemas/schemas.h"
+#include "ssx/semaphore.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/sharded.hh>
+
+#include <boost/multi_index/hashed_index.hpp>
+#include <boost/multi_index/indexed_by.hpp>
+#include <boost/multi_index/key.hpp>
+#include <boost/multi_index/member.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index/sequenced_index.hpp>
+#include <boost/multi_index_container.hpp>
+
+using namespace std::chrono_literals;
+
+namespace kafka {
+
+template<typename T>
+concept InheritsFromOCSFBase
+  = std::is_base_of<security::audit::ocsf_base_event<T>, T>::value;
+
+class audit_sink;
+
+class audit_log_manager
+  : public ss::peering_sharded_service<audit_log_manager> {
+public:
+    audit_log_manager(
+      cluster::controller* controller, kafka::client::configuration&);
+
+    audit_log_manager(const audit_log_manager&) = delete;
+    audit_log_manager& operator=(const audit_log_manager&) = delete;
+    audit_log_manager(audit_log_manager&&) = delete;
+    audit_log_manager& operator=(audit_log_manager&&) = delete;
+    ~audit_log_manager();
+
+    /// Start the underlying kafka client and create the audit log topic
+    /// if necessary
+    ss::future<> start();
+
+    /// Shuts down the internal kafka client and stops all pending bg work
+    ss::future<> stop();
+
+    /// Enqueue an event to be produced onto an audit log partition
+    ///
+    /// Returns: bool representing if the audit msg was successfully moved into
+    /// the queue or not. If unsuccessful this means the audit subsystem cannot
+    /// publish messages. Consumers of this API should react accordingly, i.e.
+    /// return an error to the client.
+    template<InheritsFromOCSFBase T, typename... Args>
+    bool enqueue_audit_event(audit_event_type type, Args&&... args) {
+        if (_as.abort_requested()) {
+            /// Prevent auditing new messages when shutdown starts that way the
+            /// queue may be entirely flushed before shutdown
+            return false;
+        }
+        if (!_auditing_enabled() || !is_audit_event_enabled(type)) {
+            return true;
+        }
+        return do_enqueue_audit_event(
+          std::make_unique<T>(std::forward<Args>(args)...));
+    }
```

Comment:
We can come back to this in a future review, but this requires the caller being exposed to the schemas in `security/audit/schemas`.  I think it would be nice to not expose those to the caller and instead have this function create them.  But this is more of a design opinion.